[Vishandi Rudy Keneta] iP

vishandi opened the pull request:
## Duke 
### By Vishandi Rudy Keneta

> "If you're trying to create a company, it's like baking a cake. You have to have all the ingredients in the right proportion." 
> -Elon Musk [source](https://www.brainyquote.com/quotes/elon_musk_567286) 

In this newly developed task-manager application, you can:
- Add tasks,
- Modify them,
- and find your task list! 

Well, it may not sound that exciting, but you know what? It's
- text-based
- easy to learn
- _~~EASY~~_ _SUPER EASY_ to use!

To feel like you have your own _Jarvis_ that manages your tasks, all you need to do is:

1. Download it from [here](https://github.com/vishandi/ip/releases)
2. double-click it.
3. add your tasks.
4. That's IT! 😉

And it's completely **FREE**!!!!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (Coming really soon in the next release!)
- [ ] Reminders (Coming really soon in the next release!)

If you are a developer yourself, you can use it as references for inspiration! Here's some sample code on `Duke.java` for you to see how beautiful it is:
```java
    public static void main(String[] args) throws DukeException {
        Duke duke = new Duke("./data/", "TaskList.txt");
        duke.initialize();
        duke.run();
    }
```

Commit:
Add Gradle support

Commit:
Added greet, echo, and exit (bye) function.

Commit:
Each text inputted is stored to a list. Added "list" function.

Commit:
Added markings of the tasks. (Level-3)

Commit:
Added Task, Event, and Todo class. Added functionality to differentiate tasks.

Commit:
fixed typo.

Commit:
A-TextUiTesting

Commit:
Added DukeException.

Commit:
Added error handling. Level-5

Commit:
changing implementation from Array to Arraylist.

Commit:
added delete command. Level-6

Commit:
fixed typo.

Commit:
Restructure TaskList

Commit:
fix bug

Commit:
add save feature

Commit:
Add dates compatibility

vishandi committed:
Merge pull request #1 from vishandi/branch-Level-7

create pull request Level-7

Commit:
resolve conflict

Commit:
fix LocalDate bug

vishandi committed:
Merge pull request #2 from vishandi/branch-Level-8

create pull request for Level-8

Commit:
Add more OOP (Ui, Storage, Parser)

Commit:
Add all classes to duke package

Commit:
Refactor to add more packages

Commit:
Create IStorage interface to prepare for JUnit testing

Commit:
Add JUnit tests for Duke and TaskList

Commit:
Add JavaDoc for public methods.

Commit:
Fix bugs from previous commit

Commit:
Tweak codes to comply with Java coding standard

Commit:
Add find feature

vishandi committed:
Merge branch 'master' into branch-A-JavaDoc

vishandi committed:
Merge pull request #3 from vishandi/branch-A-JavaDoc

Create A-JavaDoc pull request

vishandi committed:
Merge branch 'master' into branch-A-CodingStandard

vishandi committed:
Merge pull request #4 from vishandi/branch-A-CodingStandard

Create pull request for A-CodingStandard

vishandi committed:
Merge branch 'master' into branch-Level-9

vishandi committed:
Merge pull request #5 from vishandi/branch-Level-9

Create Pull Request for Level-9

Commit:
Add JavaDoc to new functions and previous functions

vishandi committed:
Merge pull request #7 from vishandi/add-gradle-support

Add Gradle support

vishandi committed:
Merge pull request #8 from vishandi/A-Gradle

A gradle

Commit:
Update build.gradle

Commit:
Add GUI

Commit:
Add FXML for GUi

Commit:
Improve code quality

vishandi committed:
Merge pull request #9 from vishandi/Level-10

Create pull request for Level-10

vishandi committed:
Update README.md

DaneMarc commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,100 @@
+     * @return ArrayList\<Task\> of all tasks that match the keyword
+     * @throws DukeException
+     */
+    public ArrayList<Task> findTasksByKeyword(String userInput, ArrayList<Task> tasks) throws DukeException {
```
I like how descriptive your method and variables names are

DaneMarc commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,100 @@
+import task.Todo;
+
+import java.time.LocalDate;
+import java.util.ArrayList;
```
I like how you organize your import statements

DaneMarc commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,119 @@
+    /**
+     * Write tasks to the file.
+     * @param taskList
+     * @throws DukeException
```
Perhaps your comments could give more detailed descriptions of the params and exceptions eg. conditions for throwing an error. I've noticed the same issue in several other places too.

DaneMarc commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,170 @@
+        } catch (DukeException d) {
+             response = this.ui.getErrorMessage(d);
+             return response;
+        }
```
I like how you handle exceptions

DaneMarc commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,170 @@
+                } else {
+                    throw DukeException.DukeInvalidCommand();
+                }
+
```
Maybe add a short comment explaining why no break is needed?

DaneMarc reviewed (commented):
Very good overall. Really well-organized and easy to read. Short and descriptive naming of variables and methods. No glaring coding standard violations. Comments could be a bit more detailed though.

Commit:
Updates JavaDoc

vishandi committed:
Merge pull request #10 from vishandi/Level-10

Updates JavaDoc

Commit:
Fix typo

Commit:
Update fxml to fix version issue

Commit:
Add assert statements and enable assertions

Commit:
Change method name for finding tasks

The method name is getMatchingTasks.

As the method is getting the message of finding the matching tasks, the
name is not accurate to capture the method's functionality.

Let's change the name of this method to improve readability.

Commit:
Change method name that parses task from user input

The current name for the method is parseFromUi.

The name is not capturing the functionality of the method accurately.

Let's change the name to improve readability.

Commit:
Move constants to static attribute of the class

Currently, there are many int constants used in the methods of Duke and
String constants used in DukeException.

Having many constants not declared as static attributes reduce
readability.

Let's move those constants to the class' static attribute.

Commit:
Abstract out string version of date in Deadline.java and Event.java

Commit:
Implement Streams to find match

vishandi committed:
Merge pull request #14 from vishandi/branch-A-Assertions

Branch a assertions

vishandi committed:
Merge branch 'master' into branch-A-CodeQuality

vishandi committed:
Merge pull request #16 from vishandi/branch-A-CodeQuality

Improve code quality

vishandi committed:
Merge pull request #17 from vishandi/branch-A-Streams

Implement Streams to find match

vishandi committed:
Update README.md

vishandi committed:
Update README.md

vishandi committed:
Add C-DetectDuplicates functionality

vishandi committed:
Merge pull request #18 from vishandi/branch-BCD-Extension

Add C-DetectDuplicates functionality

vishandi committed:
Automate exit when user inputs "bye" and fix some typo

vishandi committed:
Fix typo and add credit

vishandi committed:
Improve UI by modifying Duke's conversation

vishandi committed:
Improve MainWindow and DialogBox

vishandi committed:
Fix typo

vishandi committed:
Merge pull request #20 from vishandi/A-BetterGui

A better gui

vishandi committed:
Fix JavaDoc and some typo

vishandi committed:
Add more natural dates interpretations

vishandi committed:
Add User Guide and fix typos

vishandi committed:
Merge pull request #21 from vishandi/branch-A-UserGuide

Add User Guide and fix typos

vishandi committed:
Update README.md

vishandi committed:
Update README.md

vishandi committed:
Set theme jekyll-theme-hacker
